Ah, unfortunately this fails with
```
/mnt/mesos/sandbox/sandbox/o2/sw/SOURCES/O2/6108/0/GPU/Workflow/helper/src/GPUWorkflowHelper.cxx:149:5: error: 'trk' is not a constant expression
  149 |     if constexpr (isTPCTrack(trk)) {
```
I have no really good idea how to solve this. One could use
`template <class T> isTPCTrack() ...` and then `isTPCTrack<decltype(trk)>()`, that should work safely. I would propose to add a `std::decay_t<...>` in that case, for references, ptrs, etc.

--- a/DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h
+++ b/DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h
@@ -43,7 +43,7 @@ void o2::globaltracking::RecoContainer::createTracksVariadic(T creator) const
   // which depends on the track time:
   // 1) For track types containing TimeStampWithError ts it is ts.getTimeStamp(), getTimeStampError()
   // 2) For tracks with asymmetric time uncertainty, e.g. TPC: as mean time of t0-errBwd,t+errFwd and 0.5(errBwd+errFwd), all in TPC time bins
-  // 3) For tracks whose timing is provided as RO frame: as time in \mus for RO frame start since the start of TF, half-duration of RO window and 0.
+  // 3) For tracks whose timing is provided as RO frame: as time in \mus for RO frame start since the start of TF, half-duration of RO window.
 
   auto start_time = std::chrono::high_resolution_clock::now();
   constexpr float PS2MUS = 1e-6;
@@ -84,7 +84,7 @@ void o2::globaltracking::RecoContainer::createTracksVariadic(T creator) const
       const float timeErr = 0.010f;                                                                          // assume 10 ns error FIXME
 
       auto gidx = match.getEvIdxTrack().getIndex(); // this should be corresponding ITS-TPC track
-      if (creator(tracksPool.get(gidx), {i, GTrackID::ITSTPCTOF}, timeTOFMUS, timeErr)) {
+      if (creator(tracksTPCITS[gidx.getIndex()], {i, GTrackID::ITSTPCTOF}, timeTOFMUS, timeErr)) {
         flagUsed2(i, GTrackID::TOF);
         flagUsed(gidx); // flag used ITS-TPC tracks
       }
@@ -162,3 +162,45 @@ void o2::globaltracking::RecoContainer::createTracksVariadic(T creator) const
   auto current_time = std::chrono::high_resolution_clock::now();
   LOG(INFO) << "RecoContainer::createTracks took " << std::chrono::duration_cast<std::chrono::microseconds>(current_time - start_time).count() * 1e-6 << " CPU s.";
 }
+
+template <class T>
+inline constexpr auto isITSTrack()
+{
+  return std::is_same_v<std::decay_t<T>, o2::its::TrackITS>;
+}
+
+template <class T>
+inline constexpr auto isTPCTrack()
+{
+  return std::is_same_v<std::decay_t<T>, o2::tpc::TrackTPC>;
+}
+
+template <class T>
+inline constexpr auto isTRDTrack()
+{
+  return std::is_same_v<std::decay_t<T>, o2::trd::TrackTRD>;
+}
+
+template <class T>
+inline constexpr auto isTPCTOFTrack()
+{
+  return std::is_same_v<std::decay_t<T>, o2::dataformats::TrackTPCTOF>;
+}
+
+template <class T>
+inline constexpr auto isTPCITSTrack()
+{
+  return std::is_same_v<std::decay_t<T>, o2::dataformats::TrackTPCITS>;
+}
+
+template <class T>
+inline constexpr auto isTPCTRDTOFTrack()
+{
+  return false;
+} // to be implemented
+
+template <class T>
+inline constexpr auto isITSTPCTRDTOFTrack()
+{
+  return false;
+} // to be implemented

--- a/Detectors/GlobalTracking/include/GlobalTracking/MatchTPCITS.h
+++ b/Detectors/GlobalTracking/include/GlobalTracking/MatchTPCITS.h
@@ -428,7 +428,6 @@ class MatchTPCITS
                             std::array<int, MaxLadderCand>& lad2Check) const;
   int findLaddersToCheckBOff(int ilr, int lad0, const o2::math_utils::IntervalXYf_t& trcLinPar, float errYFrac,
                              std::array<int, MatchTPCITS::MaxLadderCand>& lad2Check) const;
-
   bool prepareTPCData();
   bool prepareITSData();
   bool prepareFITData();

--- a/Detectors/GlobalTracking/src/MatchCosmics.cxx
+++ b/Detectors/GlobalTracking/src/MatchCosmics.cxx
@@ -10,6 +10,7 @@
 
 #include "GlobalTracking/MatchCosmics.h"
 #include "DataFormatsGlobalTracking/RecoContainer.h"
+#include "DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h"
 #include "GPUO2InterfaceRefit.h"
 #include "ReconstructionDataFormats/GlobalTrackAccessor.h"
 #include "DataFormatsITSMFT/CompCluster.h"
@@ -483,26 +484,26 @@ void MatchCosmics::createSeeds(const o2::globaltracking::RecoContainer& data)
 
   mSeeds.clear();
 
-  auto creator =
-    [this](const o2::track::TrackParCov& _tr, GTrackID _origID, float t0, float terr) {
-      if (std::abs(_tr.getQ2Pt()) > this->mQ2PtCutoff) {
-        return true;
-      }
-      if (_origID.getSource() == GTrackID::TPC) { // convert TPC bins to \mus
-        t0 *= this->mTPCTBinMUS;
-        terr *= this->mTPCTBinMUS;
-      } else if (_origID.getSource() == GTrackID::ITS) { // error is supplied a half-ROF duration, convert to \mus
-        t0 += 0.5 * mITSROFrameLengthMUS;                // time 0 is supplied as beginning of ROF
-        terr *= mITSROFrameLengthMUS;
-      } else {
-        terr *= this->mMatchParams->nSigmaTError;
-      }
-      terr += this->mMatchParams->timeToleranceMUS;
-      mSeeds.emplace_back(TrackSeed{_tr, {t0 - terr, t0 + terr}, _origID, MinusOne});
+  auto creator = [this](auto& _tr, GTrackID _origID, float t0, float terr) {
+    if (std::abs(_tr.getQ2Pt()) > this->mQ2PtCutoff) {
       return true;
-    };
+    }
+    if constexpr (isTPCTrack<decltype(_tr)>()) {
+      // unconstrained TPC track, with t0 = TrackTPC.getTime0+0.5*(DeltaFwd-DeltaBwd) and terr = 0.5*(DeltaFwd+DeltaBwd) in TimeBins
+      t0 *= this->mTPCTBinMUS;
+      terr *= this->mTPCTBinMUS;
+    } else if (isITSTrack<decltype(_tr)>()) {
+      t0 += 0.5 * this->mITSROFrameLengthMUS; // time 0 is supplied as beginning of ROF in \mus
+      terr *= this->mITSROFrameLengthMUS;     // error is supplied a half-ROF duration, convert to \mus
+    } else {                                  // all other tracks are provided with time and its gaussian error in \mus
+      terr *= this->mMatchParams->nSigmaTError;
+    }
+    terr += this->mMatchParams->timeToleranceMUS;
+    mSeeds.emplace_back(TrackSeed{_tr, {t0 - terr, t0 + terr}, _origID, MinusOne});
+    return true;
+  };
 
-  data.createTracksWithMatchingTimeInfo(creator);
+  data.createTracksVariadic(creator);
 
   LOG(INFO) << "collected " << mSeeds.size() << " seeds";
 }

--- a/Detectors/GlobalTracking/src/MatchTPCITS.cxx
+++ b/Detectors/GlobalTracking/src/MatchTPCITS.cxx
@@ -41,6 +41,7 @@
 #include "ReconstructionDataFormats/Vertex.h"
 #include "GlobalTracking/MatchTPCITS.h"
 #include "DataFormatsGlobalTracking/RecoContainer.h"
+#include "DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h"
 #include "DataFormatsTPC/WorkflowHelper.h"
 
 #include "ITStracking/IOUtils.h"
@@ -529,6 +530,9 @@ void MatchTPCITS::addTPCSeed(const o2::track::TrackParCov& _tr, float t0, float 
 {
   // account single TPC seed, can be from standalone TPC track or constrained track from match to TRD and/or TOF
   const float SQRT12DInv = 2. / sqrt(12.);
+  if (_tr.getX() > o2::constants::geom::XTPCInnerRef + 0.1 || std::abs(_tr.getQ2Pt()) > mMinTPCTrackPtInv) {
+    return;
+  }
   const auto& tpcOrig = mTPCTracksArray[tpcID];
   // discard tracks w/o certain number of total or innermost pads (last cluster is innermost one)
   if (tpcOrig.getNClusterReferences() < mParams->minTPCClusters) {
@@ -600,35 +604,31 @@ bool MatchTPCITS::prepareTPCData()
     mTPCTrkLabels = inp.getTPCTracksMCLabels();
   }
 
-  auto creator = [this](const o2::track::TrackParCov& _tr, GTrackID _origID, float t0, float terr) {
-    auto srcID = _origID.getSource();
-    if (srcID == GTrackID::ITS) { // we don't need ITS here
-      return true;                // we don't need TPC tracks
+  auto creator = [this](auto& trk, GTrackID gid, float time0, float terr) {
+    if constexpr (isITSTrack<decltype(trk)>()) {
+      // do nothing, ITS tracks will be processed in a direct loop over ROFs
     }
-    // make sure the track was propagated to inner TPC radius at the ref. radius
-    if (_tr.getX() > o2::constants::geom::XTPCInnerRef + 0.1 || std::abs(_tr.getQ2Pt()) > mMinTPCTrackPtInv) {
-      return true;
-    }
-    GTrackID tpcGID{};
-    if (srcID == GTrackID::TPC) {
-      // unconstrained TPC track, with t0 = TrackTPC.getTime0+0.5*(DeltaFwd-DeltaBwd) and terr = 0.5*(DeltaFwd+DeltaBwd)
-      if (this->mSkipTPCOnly) {
-        return true;
+    if constexpr (isTPCTrack<decltype(trk)>()) {
+      // unconstrained TPC track, with t0 = TrackTPC.getTime0+0.5*(DeltaFwd-DeltaBwd) and terr = 0.5*(DeltaFwd+DeltaBwd) in TimeBins
+      if (!this->mSkipTPCOnly) {
+        this->addTPCSeed(trk, this->tpcTimeBin2MUS(time0), this->tpcTimeBin2MUS(terr), gid, gid.getIndex());
       }
-      tpcGID = _origID;
-      t0 = this->tpcTimeBin2MUS(t0); // time and error are in TPC bins, convert to \mus
-      terr = this->tpcTimeBin2MUS(terr);
-    } else if (srcID == GTrackID::TPCTOF) {
-      // TPC track is contrained by TOF fit
-      tpcGID = this->mRecoCont->getTPCContributorGID(_origID);
-    } else if (srcID == GTrackID::TPCTRD) {
-      return true; // TODO RS
     }
-    addTPCSeed(_tr, t0, terr, _origID, tpcGID.getIndex());
+    if constexpr (isTPCTOFTrack<decltype(trk)>()) {
+      // TPC track constrained by TOF time, time and its error in \mus
+      this->addTPCSeed(trk, time0, terr, gid, this->mRecoCont->getTPCContributorGID(gid));
+    }
+    if constexpr (isTRDTrack<decltype(trk)>()) {
+      // TPC track constrained by TRD trigger time, time and its error in \mus
+      LOG(ERROR) << "Not ready yet for TPC-TRD tracks";
+    }
+    if constexpr (isTPCTRDTOFTrack<decltype(trk)>()) {
+      // TPC track constrained by TRD and TOF time, time and its error in \mus
+      LOG(ERROR) << "Not ready yet for TPC-TRD-TOF tracks";
+    }
     return true;
   };
-
-  inp.createTracksWithMatchingTimeInfo(creator);
+  mRecoCont->createTracksVariadic(creator);
 
   float maxTime = 0;
   int nITSROFs = mITSROFTimes.size();

--- a/Detectors/GlobalTrackingWorkflow/src/PrimaryVertexingSpec.cxx
+++ b/Detectors/GlobalTrackingWorkflow/src/PrimaryVertexingSpec.cxx
@@ -13,6 +13,7 @@
 #include <vector>
 #include <TStopwatch.h>
 #include "DataFormatsGlobalTracking/RecoContainer.h"
+#include "DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h"
 #include "ReconstructionDataFormats/TrackTPCITS.h"
 #include "ReconstructionDataFormats/GlobalTrackID.h"
 #include "DetectorsBase/Propagator.h"
@@ -110,24 +111,24 @@ void PrimaryVertexingSpec::run(ProcessingContext& pc)
   auto halfROFITS = 0.5 * mITSROFrameLengthMUS;
   auto hw2ErrITS = 2.f / std::sqrt(12.f) * mITSROFrameLengthMUS; // conversion from half-width to error for ITS
 
-  auto creator =
-    [maxTrackTimeError, hw2ErrITS, halfROFITS, &tracks, &gids](const o2::track::TrackParCov& _tr, GTrackID _origID, float t0, float terr) {
-      if (!_origID.includesDet(DetID::ITS)) {
-        return true; // just in case this selection was not done on RecoContainer filling level
-      }
-      if (_origID.getSource() == GTrackID::ITS) { // error is supplied a half-ROF duration, convert to \mus
-        t0 += halfROFITS;                         // ITS time is supplied as beginning of ROF
-        terr *= hw2ErrITS;
-      }
-      if (terr > maxTrackTimeError) {
-        return true;
-      }
+  auto creator = [maxTrackTimeError, hw2ErrITS, halfROFITS, &tracks, &gids](auto& _tr, GTrackID _origID, float t0, float terr) {
+    if (!_origID.includesDet(DetID::ITS)) {
+      return true; // just in case this selection was not done on RecoContainer filling level
+    }
+    if constexpr (isITSTrack<decltype(_tr)>()) {
+      t0 += halfROFITS;  // ITS time is supplied in \mus as beginning of ROF
+      terr *= hw2ErrITS; // error is supplied as a half-ROF duration, convert to \mus
+    }
+    // for all other tracks the time is in \mus with gaussian error
+    if (terr < maxTrackTimeError) {
       tracks.emplace_back(TrackWithTimeStamp{_tr, {t0, terr}});
       gids.emplace_back(_origID);
-      return true;
-    };
+    }
+    return true;
+  };
 
-  recoData.createTracksWithMatchingTimeInfo(creator); // create track sample considered for vertexing
+  recoData.createTracksVariadic(creator); // create track sample considered for vertexing
+
   if (mUseMC) {
     recoData.fillTrackMCLabels(gids, tracksMCInfo);
   }

--- a/Detectors/Vertexing/include/DetectorsVertexing/PVertexer.h
+++ b/Detectors/Vertexing/include/DetectorsVertexing/PVertexer.h
@@ -197,7 +197,6 @@ void PVertexer::createTracksPool(const TR& tracks, gsl::span<const o2d::GlobalTr
   mTracksPool.reserve(ntGlo);
   // check all containers
   float vtxErr2 = 0.5 * (mMeanVertex.getSigmaX2() + mMeanVertex.getSigmaY2());
-  float pullIniCut = 9.; // RS FIXME  pullIniCut should be a parameter
   o2d::DCA dca;
 
   for (uint32_t i = 0; i < ntGlo; i++) {

--- a/Detectors/Vertexing/src/VertexTrackMatcher.cxx
+++ b/Detectors/Vertexing/src/VertexTrackMatcher.cxx
@@ -12,6 +12,7 @@
 /// \brief Class for vertex track association
 /// \author ruben.shahoyan@cern.ch
 
+#include "DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h"
 #include "DetectorsVertexing/VertexTrackMatcher.h"
 #include "DataFormatsParameters/GRPObject.h"
 #include "DetectorsCommonDataFormats/NameConf.h"
@@ -151,25 +152,24 @@ void VertexTrackMatcher::extractTracks(const o2::globaltracking::RecoContainer& 
 
   mTBrackets.clear();
 
-  auto creator =
-    [this, &vcont](const o2::track::TrackParCov& _tr, GIndex _origID, float t0, float terr) {
-      if (vcont.find(_origID) != vcont.end()) { // track is contributor to vertex, already accounted
-        return true;
-      }
-      if (_origID.getSource() == GIndex::TPC) { // convert TPC bins to \mus
-        t0 *= this->mTPCBin2MUS;
-        terr *= this->mTPCBin2MUS;
-      } else if (_origID.getSource() == GIndex::ITS) { // error is supplied a half-ROF duration, convert to \mus
-        t0 += this->mITSROFrameLengthMUS;
-        terr *= 0.5 * this->mITSROFrameLengthMUS;
-      } else {
-        //terr *= this->mMatchParams->nSigmaTError;
-      }
-      mTBrackets.emplace_back(TrackTBracket{{t0 - terr, t0 + terr}, _origID});
+  auto creator = [this, &vcont](auto& _tr, GIndex _origID, float t0, float terr) {
+    if (vcont.find(_origID) != vcont.end()) { // track is contributor to vertex, already accounted
       return true;
-    };
+    }
+    if constexpr (isTPCTrack<decltype(_tr)>()) {
+      // unconstrained TPC track, with t0 = TrackTPC.getTime0+0.5*(DeltaFwd-DeltaBwd) and terr = 0.5*(DeltaFwd+DeltaBwd) in TimeBins
+      t0 *= this->mTPCBin2MUS;
+      terr *= this->mTPCBin2MUS;
+    } else if (isITSTrack<decltype(_tr)>()) {   // error is supplied a half-ROF duration, convert to \mus
+      t0 += this->mITSROFrameLengthMUS;         // ITS time is supplied in \mus as beginning of ROF
+      terr *= 0.5 * this->mITSROFrameLengthMUS; // error is supplied as a half-ROF duration, convert to \mus
+    }
+    // for all other tracks the time is in \mus with gaussian error
+    mTBrackets.emplace_back(TrackTBracket{{t0 - terr, t0 + terr}, _origID});
+    return true;
+  };
 
-  data.createTracksWithMatchingTimeInfo(creator);
+  data.createTracksVariadic(creator);
 
   // sort in increasing min.time
   std::sort(mTBrackets.begin(), mTBrackets.end(), [](const TrackTBracket& a, const TrackTBracket& b) { return a.tBracket.getMin() < b.tBracket.getMin(); });

--- a/GPU/Workflow/helper/src/GPUWorkflowHelper.cxx
+++ b/GPU/Workflow/helper/src/GPUWorkflowHelper.cxx
@@ -146,7 +146,7 @@ std::shared_ptr<const GPUWorkflowHelper::tmpDataContainer> GPUWorkflowHelper::fi
   }
 
   auto creator = [&recoCont, &retVal](auto& trk, GID gid, float time, float) {
-    if constexpr (std::is_same_v<std::decay_t<decltype(trk)>, o2::tpc::TrackTPC>) {
+    if constexpr (isTPCTrack<decltype(trk)>()) {
       time = trk.getTime0();
     }
     retVal->globalTracks.emplace_back(&trk);

